channel/saver.go

```diff
@@ -0,0 +1,329 @@
+package channel
+
+import (
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Hucaru/Valhalla/common"
+)
+
+// DirtyBits mark which character columns need persisting.
+type DirtyBits uint32
+
+const (
+	DirtyAP DirtyBits = 1 << iota
+	DirtySP
+	DirtyMesos
+	DirtyHP
+	DirtyMP
+	DirtyMaxHP
+	DirtyMaxMP
+	DirtyEXP
+	DirtyMap
+	DirtyJob
+	DirtyLevel
+	DirtyStr
+	DirtyDex
+	DirtyInt
+	DirtyLuk
+	DirtyFame
+)
+
+// snapshot contains only columns we may persist.
+type snapshot struct {
+	ID int32
+
+	AP, SP         int16
+	Mesos          int32
+	HP, MaxHP      int16
+	MP, MaxMP      int16
+	EXP            int32
+	MapID          int32
+	MapPos         byte
+	Job            int16
+	Level          byte
+	Str, Dex, Intt int16
+	Luk, Fame      int16
+}
+
+// pendingSave is a coalesced, debounced update for one character.
+type pendingSave struct {
+	due  time.Time
+	bits DirtyBits
+	snap snapshot
+}
+
+// Saver batches and persists character updates.
+type Saver struct {
+	mu      sync.Mutex
+	pending map[int32]*pendingSave // charID -> pendingSave
+	ticker  *time.Ticker
+	stopCh  chan struct{}
+}
+
+// SaverInstance is the global saver used by the channel server.
+var SaverInstance *Saver
```

Comment:
I think a better singleton pattern would be not export this variable or the type and use functions to perform the work